src/deepsparse/pipeline.py

```diff
@@ -176,22 +174,113 @@ def __init__(
         self.onnx_file_path = self.setup_onnx_file_path()
         self.engine = self._initialize_engine()
 
-    def __call__(self, *args, **kwargs) -> Union[BaseModel, Future]:
-        _default_key_val = ("_DEFAULT",)
-        executor = kwargs.get("executor", _default_key_val)
+        self._batch_size = self._batch_size or 1
 
-        if executor is _default_key_val:  # do not use ==
-            # use executor created during initialization
-            executor = self.executor
+    def __call__(self, *args, **kwargs) -> BaseModel:
+        if "engine_inputs" in kwargs:
+            raise ValueError(
+                "invalid kwarg engine_inputs. engine inputs determined "
+                f"by {self.__class__.__qualname__}.parse_inputs"
+            )
+
+        # parse inputs into input_schema
+        pipeline_inputs = self.parse_inputs(*args, **kwargs)
+        if not isinstance(pipeline_inputs, self.input_schema):
+            raise RuntimeError(
+                f"Unable to parse {self.__class__} inputs into a "
+                f"{self.input_schema} object. Inputs parsed to {type(pipeline_inputs)}"
+            )
+
+        # batch size of the inputs may be `> self._batch_size` at this point
+        engine_inputs: List[numpy.ndarray] = self.process_inputs(pipeline_inputs)
+        if isinstance(engine_inputs, tuple):
+            engine_inputs, postprocess_kwargs = engine_inputs
         else:
-            # use passed in executor
-            executor = kwargs.pop("executor")
+            postprocess_kwargs = {}
+
+        # split inputs into batches of size `self._batch_size`
+        batches = self.split_engine_inputs(engine_inputs, self._batch_size)
+
+        # submit to engine
+        futures = [
+            self.executor.submit(self.engine_forward, batch) for batch in batches
+        ]
+        wait(futures)
+
+        # join together the batches of size `self._batch_size`
+        engine_outputs = self.join_engine_outputs(
+            [future.result() for future in futures]
+        )
```

Comment:
awesome, pretty slick. could we do this for the base `Engine` class so we can support dynamic batch outside of pipelines?

Reply:
Yeah I don't see why not! We are planning on using the executor in subclasses of pipelines to enable asynchronous pre/postprocessing. So if we did that we'd need to make sure all the classes are using the same executor instance.